Add datalab notebook

elisno opened the pull request:
This notebook focuses on running Datalab with a subset of the Caltech-256 dataset. It uses precomputed pred_probs and features.

Remaining todos:

- [x] Move features and pred_probs to S3.
- [x] Add notebook that generates features + pred_probs and saves them. See https://github.com/cleanlab/examples/pull/50#issuecomment-1513548981.

elisno committed:
add datalab notebook

jwmueller commented:
Can you add a second notebook that shows how to generate the pred_probs and features + save them to the files that are then loaded?

elisno commented:
> Can you add a second notebook that shows how to generate the pred_probs and features + save them to the files that are then loaded?

Ah yes, thanks! Added todo in the description.

jwmueller commented on datalab/requirements-datalab.txt:
```diff
@@ -0,0 +1,3 @@
+git+https://github.com/elisno/cleanlab.git@datalab#egg=cleanlab[datalab]
+matplotlib
+ipykernel
```
why is ipykernel required?

elisno replied:
Just for local environment. Removed.

elisno committed:
feat: add notebook for training transformer on caltech-256 subset

Include separate requirements file. Add README.md. Simplify filenames of pre-computed feature and pred_probs. Fetch those also from S3 bucket.

huiwengoh commented:
[nit] I think we can remove the `_subset` suffix from the titles of the notebook, it feels a little distracting imo. 
The main one can just be `datalab.ipynb`, perhaps the other can be something like `train_caltech.ipynb`?

jwmueller commented on datalab/train_subset.ipynb:
```diff
@@ -0,0 +1,501 @@
+{
```
Rename this notebook: train_image_classifier.ipynb

jwmueller commented on datalab/train_subset.ipynb:
```diff
@@ -0,0 +1,501 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Dataset Preparation\n",
```
If it's easy to do, I'd skip this entire data prep section and just load the prepped dataset from:

```
!wget -nc 'https://cleanlab-public.s3.amazonaws.com/Datalab/caltech256-subset.tar.gz'
```

This dataset prep section doesn't really help somebody better understand how to train an image classifier on their own data which is the purpose of this notebook

jwmueller replied:
If not easy to change though, then just keep as is.

elisno replied:
Easy enough.

jwmueller commented on datalab/train_subset.ipynb:
```diff
@@ -0,0 +1,501 @@
+   "source": [
+    "## Feature Extraction\n",
+    "\n",
+    "After training the model, we will compute predicted probabilities for the entire dataset using the trained models from each fold.\n",
```
why are you discussing predicted probabilities in this section?

elisno replied:
Renamed section. The previous section was only related to training the models (with early stopping).

`pred_probs` and `features` are both computed in this section.

jwmueller commented on datalab/train_subset.ipynb:
```diff
@@ -0,0 +1,501 @@
+    "\n",
+    "In this notebook, we will train a Swin Transformer model for image classifier on a subset of the Caltech-256 dataset using the Timm library and Pytorch. The main steps include dataset sampling and precrocessing, model training with k-fold cross-validation, and feature extraction.\n"
+   ]
+  },
```
clarify somewhere that Datalab does not actually require both features and pred_probs, but providing both allows it to more comprehensively detect issues in a dataset.

jwmueller commented on datalab/train_subset.ipynb:
```diff
@@ -0,0 +1,501 @@
+    "\n",
+    "During the training process, we set aside a validation split on a per-fold basis to allow early stopping for each fold. This approach helps us prevent overfitting and obtain a better estimate of the model's performance.\n",
+    "\n",
+    "While we're not specifically interested in the model artifacts themselves, we aim to get a general idea of whether the chosen model architecture is accurate enough for our purpose."
```
warn that this cell may take a long time to execute and should be run with a GPU

elisno committed:
address comments on training notebook

elisno committed:
address datalab notebook feedback, simplify variable names, split sections on filtering dataframe and plotting images

jwmueller commented on datalab/README.md:
```diff
@@ -0,0 +1,18 @@
+# Training a Swin Transformer model with PyTorch on a subset of Caltech-256 for auditing with Datalab
+This example demonstrates the use of the following module below from cleanlab:
```
```suggestion
This example demonstrates cleanlab's Datalab class for auditing an image dataset.
```

jwmueller commented on datalab/README.md:
````diff
@@ -0,0 +1,18 @@
+    pip install -r requirements-train.txt
+    ```
+
+- [datalab.ipynb](datalab.ipynb) - Runs the dataset through Datalab with the trained model artifacts.
+  - Install dependencies with
````
```suggestion
- [datalab.ipynb](datalab.ipynb) - Audits the dataset using Datalab applied to outputs from the trained model.
  - Install dependencies with
```

jwmueller commented on datalab/README.md:
```diff
@@ -0,0 +1,18 @@
+This example demonstrates the use of the following module below from cleanlab:
+
+It is split into two notebooks:
+- [train_image_classifier.ipynb](train_image_classifier.ipynb) - Trains a Swin Transformer model on a subset of Caltech-256
```
```suggestion
- [train_image_classifier.ipynb](train_image_classifier.ipynb) - Trains a Swin Transformer classifier model on a subset of Caltech-256
```

jwmueller commented on datalab/README.md:
```diff
@@ -0,0 +1,18 @@
+# Training a Swin Transformer model with PyTorch on a subset of Caltech-256 for auditing with Datalab
```
```suggestion
# Detecting issues in an image classification dataset (Caltech-256) with Datalab and a Swin Transformer model.
```

jwmueller commented on datalab/requirements-datalab.txt:
```diff
@@ -0,0 +1,2 @@
+git+https://github.com/elisno/cleanlab.git@datalab#egg=cleanlab[datalab]
```
don't forget to update this when next version of cleanlab is released that contains Datalab

elisno replied:
Switching to cleanlab/cleanlab repo for now, but keeping this reminder for release.

jwmueller commented on datalab/requirements-train.txt:
```diff
@@ -0,0 +1,9 @@
+scikit-learn
+timm
+tqdm
+--extra-index-url https://download.pytorch.org/whl/cu116
```
is this really something to specify for an arbitrary user?

elisno replied:
Moved this to the readme as a pip install instruction like:

```
python -m pip install --extra-index-url http://my.package.repo/simple SomePackage
```

jwmueller committed:
new readme title

jwmueller committed:
correct language

jwmueller committed:
wording

jwmueller commented on datalab/README.md:
```diff
@@ -0,0 +1,18 @@
+# Detecting issues in an image classification dataset (Caltech-256) with Datalab and a Swin Transformer model.
```
Add a reference to this example in the table of contents from the main readme in examples/

jwmueller committed:
add datalab example to TOC

jwmueller committed:
shorten description

jwmueller commented on datalab/README.md:
```diff
@@ -0,0 +1,18 @@
+# Detecting issues in an image classification dataset (Caltech-256) with Datalab and a Swin Transformer model.
+This example demonstrates cleanlab's Datalab class for auditing an image dataset.
```
```suggestion
# Detecting issues in an image classification dataset (Caltech-256) with Datalab and a Swin Transformer model.

This example demonstrates cleanlab's Datalab class for auditing an image dataset.
```

jwmueller committed:
spacing

jwmueller commented on README.md:
```diff
@@ -8,6 +8,7 @@ To quickly learn how to run cleanlab on your own data, first check out the [quic
 
 |     | Example                                                                                        | Description                                                                                                                                                                                                                                                                  |
 | --- | ---------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| 0   | [datalab](datalab_image_classification)                              | Use Datalab to detect issues in (a subset of) the Caltech-256 image classification dataset.                        |
```
TODO: @elisno 

Please rename the entire folder where all these files live to be: `datalab_image_classification` as I've specified in this table of contents link here.

jwmueller committed:
polish training notebook

jwmueller committed:
jupyter formatting

jwmueller committed:
remove extra comma

jwmueller committed:
wording

jwmueller committed:
rename folder

jwmueller committed:
editing notebooks

elisno committed:
run examples notebook with latest code changes

elisno committed:
update readme

elisno committed:
move extra url for torch package index to pip install instructions

huiwengoh committed:
minor edits

huiwengoh reviewed (approved):
One small question/comment: should we pin some of the dependency versions in `requirements.txt` ?

Otherwise LGTM!

elisno committed:
build: ⬆️ switch cleanlab[datalab] dependency to master branch of main repo

jwmueller commented on datalab_image_classification/README.md:
```diff
@@ -0,0 +1,19 @@
+# Detecting issues in an image classification dataset (Caltech-256) with Datalab and a Swin Transformer model.
+
+This example demonstrates cleanlab's Datalab class for auditing an image dataset.
+
+It is split into two notebooks:
```
```suggestion
# Detecting issues in an image classification dataset (Caltech-256) with Datalab

This example uses cleanlab's Datalab class to audit an image dataset. Here we run Datalab with a Swin Transformer model trained for classification.

There are two notebooks:
```

jwmueller committed:
wording

jwmueller reviewed (approved):
LGTM!

elisno committed:
build: 📌 pin dependencies for caltech256-subset training notebook
